### PR TITLE
WIP: make internal methods private and move helpers to utils.py

### PR DESCRIPTION
**Pull request by pavanramkumar**

### CODE REVIEW

**Comment by pavanramkumar** [comment]
note that `grad_L2loss` is still public because `test_pyglmnet.py` calls it

**Comment by pavanramkumar** [comment]
closes #50

File: pyglmnet/pyglmnet.py [hunk 1]
```diff
@@ -232,7 +169,7 @@ def copy(self):
         """Return a copy of the object."""
         return deepcopy(self)
 
-    def qu(self, z):
+    def __qu(self, z):
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: do I see a double underscore? @pavanramkumar I think you need only a single underscore

File: pyglmnet/__init__.py [hunk 2]
```diff
@@ -1,2 +1,3 @@
-from .pyglmnet import GLM, set_log_level, label_binarizer
+from pyglmnet import GLM, set_log_level
+from utils import softmax, label_binarizer, log_likelihood
```
Comment (jasmainak) [inline pyglmnet/__init__.py]: from `.utils`, no?

File: pyglmnet/pyglmnet.py [hunk 3]
```diff
@@ -5,6 +5,7 @@
 
 import numpy as np
 from scipy.special import expit
+import utils as utils
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: you should use relative imports for importing functions, otherwise people might end up importing this: https://pypi.python.org/pypi/utils

File: examples/plot_poisson_2.py [hunk 4]
```diff
@@ -90,7 +91,14 @@
 ########################################################
 
 z = np.linspace(0., 10., 100)
-qu = glm_poissonexp.qu(z)
+
+eta = 4.0
+qu = deepcopy(z)
```
Comment (jasmainak) [inline examples/plot_poisson_2.py]: maybe stupid question but why is `qu` always used as an abbreviation for "conditional intensity" ? same question for `lmb`. I couldn't figure it out yet

File: pyglmnet/utils.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    maxes = maxes.reshape(maxes.shape[0], 1)
+    e = np.exp(w - maxes)
+    dist = e / np.sum(e, axis=1, keepdims=True)
+    return dist
+
+
+def label_binarizer(y):
+    """Mimics scikit learn's LabelBinarizer
+
+
```
Comment (jasmainak) [inline pyglmnet/utils.py]: bad extra line

File: pyglmnet/utils.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    Parameters
+    ----------
+    y: ndarray (n_samples)
```
Comment (jasmainak) [inline pyglmnet/utils.py]: you mean the shape is `(n_samples)`? Two comments here: you need to specify that it's the shape. Also, shape is a tuple, so you need to say `(n_samples, )`. Compare:

```py
>>> x = (100)
>>> type(x)
```
and 

```py
>>> x = (100, )
>>> type(x)
```

File: tests/test_pyglmnet.py [hunk 7]
```diff
@@ -95,9 +95,9 @@ def test_multinomial():
                  tol=1e-10)
     X = np.array([[-1, -2, -3], [4, 5, 6]])
     y = np.array([1, 0])
+
```
Comment (jasmainak) [inline tests/test_pyglmnet.py]: it's usually not considered good practice to touch parts of code not related to the PR. Maybe I'm being a bit too nitpicky ;)

File: pyglmnet/utils.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    """
+    if y.ndim != 1:
+        raise ValueError('y has to be one-dimensional')
+    y_flat = y.ravel()
+    yb = np.zeros([len(y), y.max() + 1])
+    yb[np.arange(len(y)), y_flat] = 1
+    return yb
+
+
+def log_likelihood(y, yhat, distr):
```
Comment (jasmainak) [inline pyglmnet/utils.py]: shouldn't this be private too?

File: pyglmnet/utils.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
+import numpy as np
```
Comment (jasmainak) [inline pyglmnet/utils.py]: can you add a minimal docstring, authors and the license to the file? Something like this: https://github.com/nilearn/nilearn/blob/master/nilearn/image/image.py

**Comment by jasmainak** [comment]
> note that grad_L2loss is still public because test_pyglmnet.py calls it

you _can_ in principle still use a private method / function by doing `object._private_method()`. Although, the fact that it is private means that the developers could change the API at any time without a deprecation. Keeping things private means it gives developers the flexibility to change the structure of the code as things evolve ...

So, I would say it's okay to make `grad_L2loss` private but use it in the tests as long as we don't do it too often.

File: pyglmnet/pyglmnet.py (hunk 1 above)
Comment (pavanramkumar) [inline pyglmnet/pyglmnet.py]: @jasmainak i followed these conventions: http://www.diveintopython.net/object_oriented_framework/private_functions.html

if you use a double underscore, you will note be able to call `object.__private_method()`

Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: woops, I wish I had finished that book ;) python has so many features I'm learning all the time. Anyway, I found this on stack overflow: http://stackoverflow.com/a/6930223. Seems like there is a good reason everyone uses a single underscore

Comment (pavanramkumar) [inline pyglmnet/pyglmnet.py]: ok, makes sense :) this way i can make `grad_L2loss` to be pseudo-private

File: pyglmnet/utils.py (hunk 5 above)
Comment (pavanramkumar) [inline pyglmnet/utils.py]: flake8 expects two line spaces, otherwise travis build fails :)

File: tests/test_pyglmnet.py (hunk 7 above)
Comment (pavanramkumar) [inline tests/test_pyglmnet.py]: again, flake8 complained about spaces :)

File: pyglmnet/utils.py (hunk 5 above)
Comment (jasmainak) [inline pyglmnet/utils.py]: I'm puzzled. Why does it work for the `softmax` function then? You have only one blank line there, no?

File: pyglmnet/utils.py (hunk 8 above)
Comment (pavanramkumar) [inline pyglmnet/utils.py]: hmm, it's not a class method? it's just a helper function: so same category as `softmax()` or `label_binarizer()`

Comment (jasmainak) [inline pyglmnet/utils.py]: private if you don't expect users to use it :) I thought `label_binarizer` was something users were going to use, no?

Comment (pavanramkumar) [inline pyglmnet/utils.py]: i would let the core `GLM` methods alone be part of the API. 

we created our own `softmax()` and `label_binarizer()` to avoid `scikit-learn` dependency.
